### ISSUE TITLE
[QST] Support for seq2seq tasks?

### ISSUE THREAD

**Issue by DiTo97** [NVIDIA-Merlin/Transformers4Rec#601]
# ❓ Questions & Help

## Details

Does the T4R package support sequence-to-sequence tasks?

For instance, let's say we are working with e-commerce data: given a truncated user session, we are not only interested in the next item the user will interact with, but rather in the ordered sequence of $k$ interactions he/she will have. Is there any way to accomplish this with the package in its current state?

**Comment by Rohan0401** [comment]
Yes it can be done. 
Suppose you have a trainer instance 
`# Instantiate the T4Rec Trainer, which manages training and evaluation
trainer = Trainer(
    model=model,
    args=training_args,
    schema=schema,
    compute_metrics=True,
)`
You can get the top k prediction using 
test_paths=<file path>
trainer.args.predict_top_k = 20 # for top 20 recommendations 
prediction = trainer.predict(test_paths)
print(prediction.predictions)

Hope it helps.

**Comment by rnyak** [comment]
@DiTo97 if your goal is to generate the topK predictions for a given user session, you can do that as Rohan mentioned above. 
```
test_paths=<set the path to your test set>
trainer.args.predict_top_k = 20  # you can change this value
prediction = trainer.predict(test_paths)
print(prediction)
```

**Comment by DiTo97** [comment]
> @DiTo97 if your goal is to generate the topK predictions for a given user session, you can do that as Rohan mentioned above.
> 
> ```
> test_paths=<set the path to your test set>
> trainer.args.predict_top_k = 20  # you can change this value
> prediction = trainer.predict(test_paths)
> print(prediction)
> ```

Hi @rnyak and @Rohan0401,

If I used the `top_k` parameter, say $k = 20$ would I get the top $k$ predictions at the next step or over the next $k$ steps?

I will make an example to better clarify my concern. Let's say that I have a user session from $0$ to $t$, would I get the top $k$ possibilities at $t + 1$ or the best sequence of predictions from $t + 1$ to $t + k$? I'm looking for the latter.

I suppose that if the `top_k` parameter covers the former case, i.e., the top $k$ possibilities at $t + 1$, I may iterate the next item prediction process $k$ times, always feeding it with the previous predicted next items, to decode a sequence up to $t + k$ (implementation details of the decoding strategy are negligible at this stage, i.e., beam search with beam size $n$, or similar).

Rather, my question was if a head for such a task was already supported by T4R, for training and inference.

**Comment by gabrielspmoreira** [comment]
Hey @DiTo97 . Thanks for the question. We currently don't have a decoder head. We use Transformers blocks as encoders, where you get a representation for each position of the input sequence. 
But you could build such seq-2-seq decoder yourself on top of our encoders using HF Transformer layers and a custom loss function. You would have more control by having your own train loop instead of using Trainer.train().
Also, your proposed approach to generate the next k steps iteratively by combining next-item prediction with beam search and extending the original sequence with predicted ones would work. But the model would be still trained for optimizing next-item prediction and not the whole sequence continuation.

**Comment by DiTo97** [comment]
Hey @gabrielspmoreira,

Thank you for the clarification. I have already started implementing the first approach that you mentioned above and I think it is the way to go to extend the capabilities of the Transformer encoders currently implemented in T4R.

I'm closing the issue!

**Comment by jian-mo-proton** [comment]
Hi @DiTo97, I'm working on extending the transformer4rec for seq2seq too. Could you share your work on this? Thanks a lot.
